File: CpogsPlugin/src/org/workcraft/plugins/cpog/CPOG.java
/*
 *
 * Copyright 2008,2009 Newcastle University
 *
 * This file is part of Workcraft.
 * 
 * Workcraft is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Workcraft is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Workcraft.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package org.workcraft.plugins.cpog;

import static org.workcraft.dependencymanager.advanced.core.GlobalCache.*;

import org.workcraft.dependencymanager.advanced.core.EvaluationContext;
import org.workcraft.dependencymanager.advanced.core.Expression;
import org.workcraft.dependencymanager.advanced.core.ExpressionBase;
import org.workcraft.dependencymanager.advanced.user.ModifiableExpression;
import org.workcraft.dependencymanager.advanced.user.StorageManager;
import org.workcraft.util.Function;
import org.workcraft.util.Function0;

import pcollections.PVector;
import pcollections.TreePVector;

public class CPOG
{
	public final StorageManager storage;
	
	Function0<String> varNameGen = new CheckedPrefixNameGen("x_", new Function<String, Boolean>(){
		@Override
		public Boolean apply(String candidate) {
			for(Variable var : eval(variables))
				if(eval(var.label).equals(candidate))
					return false;
			return true;
		}
	});
	
	Function0<String> vertNameGen = new CheckedPrefixNameGen("v_", new Function<String, Boolean>(){
		@Override
		public Boolean apply(String candidate) {
			for(Vertex v : eval(vertices))
				if(eval(v.visualInfo.label).equals(candidate))
					return false;
			return true;
		}
	});
	
	final ModifiableExpression<PVector<Variable>> variables;
	final ModifiableExpression<PVector<Vertex>> vertices;
	final ModifiableExpression<PVector<RhoClause>> rhoClauses;
	final ModifiableExpression<PVector<Arc>> arcs;
	final ModifiableExpression<PVector<DynamicVariableConnection>> dynamicArcs;

	<T>ModifiableExpression<PVector<T>> createEmpty(StorageManager storage) {
		PVector<T> empty = TreePVector.empty();
		return storage.create(empty);
	}
	
	public CPOG(StorageManager storage) {
		this.storage = storage;
		variables = createEmpty(storage);
		vertices = createEmpty(storage);
		rhoClauses = createEmpty(storage);
		arcs = createEmpty(storage);
		dynamicArcs = createEmpty(storage);
	}
	
	static <T> void add(ModifiableExpression<PVector<T>> vec, T item) {
		vec.setValue(eval(vec).plus(item));
	}

	public Arc connect(Vertex first, Vertex second) {
		Arc con = new Arc(first, second, storage);
		add(arcs, con);
		return con;
	}

	public Vertex createVertex() {
		Vertex vertex = Vertex.make(storage);
		vertex.visualInfo.label.setValue(vertNameGen.apply());
		add(vertices, vertex);
		return vertex;
	}

	public RhoClause createRhoClause() {
		RhoClause rhoClause = RhoClause.make(storage);
		add(rhoClauses, rhoClause);
		return rhoClause;
	}

	public Variable createVariable() {
		String name = varNameGen.apply();
		Variable var = Variable.make(storage, storage.create(name));
		add(variables, var);
		return var;
	}

	public Expression<PVector<Node>> nodes() {
		return new ExpressionBase<PVector<Node>>(){

			@Override
			protected PVector<Node> evaluate(EvaluationContext context) {
				return TreePVector.<Node>empty()
					.plusAll(context.resolve(variables));
			}
			
		};
	}
}

File: CpogsPlugin/src/main/scala/org/workcraft/plugins/cpog/scala/CPOG.scala
package org.workcraft.plugins.cpog.scala
import org.workcraft.util.Function0
import org.workcraft.plugins.cpog.CheckedPrefixNameGen
import org.workcraft.scala.Util._
import org.workcraft.scala.Scalaz._
import org.workcraft.scala.Expressions._
import org.workcraft.scala.StorageManager
import org.workcraft.plugins.cpog.scala.nodes._

object CPOG {
  def apply(storage : StorageManager) : CPOG = apply(storage, Nil, Nil, Nil, Nil)
  def apply(storage : StorageManager, 
      variables : List[Variable], 
      vertices : List[Vertex], 
      rhoClauses : List[RhoClause],
      arcs : List[Arc]) = new CPOG(storage, 
          storage.create(variables), 
          storage.create(vertices),
          storage.create(rhoClauses),
          storage.create(arcs))
}

class CPOG(val storage : StorageManager,
    val variables : ModifiableExpression[List[org.workcraft.plugins.cpog.scala.nodes.Variable]],
    val vertices : ModifiableExpression[List[org.workcraft.plugins.cpog.scala.nodes.Vertex]],
    val rhoClauses : ModifiableExpression[List[org.workcraft.plugins.cpog.scala.nodes.RhoClause]],
    val arcs : ModifiableExpression[List[org.workcraft.plugins.cpog.scala.nodes.Arc]]
  ) {

  val varNameGen : Function0[String] = new CheckedPrefixNameGen("x_", (candidate : java.lang.String) => {
			eval(variables).foldLeft(true)((b, v) => b && !eval(v.visualProperties.label).equals(candidate))
		} : java.lang.Boolean
	)
	
  val vertNameGen : Function0[String] = new CheckedPrefixNameGen("v_", (candidate : java.lang.String) => {
    val v : List[Vertex] = eval(vertices)
    def ff(b : java.lang.Boolean,v : Vertex) : java.lang.Boolean = b && !eval(v.visualProperties.label).equals(candidate)
    v.foldLeft(true : java.lang.Boolean)(ff)
  })

  def add[T](vec : ModifiableExpression[List[T]], item : T) {
    vec.modify(item :: _)
  }

  def connect(first : Vertex, second : Vertex) : Arc = {
    val con : Arc = Arc.create(storage, first, second)
    add(arcs, con)
    con
  }

  def createVertex : Vertex = {
    val vertex = Vertex.create(storage)
    vertex.visualProperties.label.setValue(vertNameGen.apply)
    add(vertices, vertex);
    return vertex;
  }

  def createRhoClause : RhoClause = {
    val rhoClause = RhoClause.create(storage)
    add(rhoClauses, rhoClause)
    return rhoClause
  }

  def createVariable : Variable = {
    val name = varNameGen.apply
    val v = Variable.create(storage, storage.create(name))
    add(variables, v);
    return v;
  }

  val components : Expression[List[Component]] = {
    for(variables <- variables;
      vertices <- vertices;
      rhoClauses <- rhoClauses)
    yield variables ::: vertices ::: rhoClauses
  }

  val nodes : Expression[List[Node]] = {
    for(c <- components;
    a <- arcs)
      yield c ::: a
  }
}